rust/lance-index/benches/hnsw.rs

```diff
@@ -31,7 +31,7 @@ fn bench_hnsw(c: &mut Criterion) {
 
     let data = generate_random_array_with_seed::<Float32Type>(TOTAL * DIMENSION, SEED);
     let fsl = FixedSizeListArray::try_new_from_values(data, DIMENSION as i32).unwrap();
-    let vectors = Arc::new(FlatStorage::new(fsl.clone(), DistanceType::L2));
+    let vectors = Arc::new(FlatFloatStorage::new(fsl.clone(), DistanceType::L2));
```

Comment:
Ah, do we need FlatStorage to be typed? This is sad

Reply:
yeah... unfortunately now we have to do this.
there are 2 ways to avoid this in my mind but they would introduce some new issues as well:
1. storing `ArrayRef` in dist calculator then cast it to `&PrimitiveArray<T>` according distance type / column type. This is what we did before, it's slow
2. just store everything in bytes ref `&[u8]` then do unsafe cast